### PR TITLE
Convert rib-compiler-app to Kotlin

### PR DESCRIPTION
**Pull request by jbarr21**
Migrates the `rib-compiler-app` module to Kotlin as part of #412

### CODE REVIEW

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/AnnotationVerifier.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * limitations under the License.
+ */
+package com.uber.rib.compiler
+
+import javax.lang.model.element.TypeElement
+import javax.lang.model.util.Elements
+import javax.lang.model.util.Types
+
+/**
+ * Verifier that checks if the type element is valid and returns information as [ ].
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/AnnotationVerifier.kt:23]: nit - should this be "returns information as `[AnnotatedClass]`?"

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  private fun hasConstructor(type: TypeElement): Boolean {
+    val constructors: MutableList<ExecutableElement> = ArrayList()
+    for (element in elementUtils.getAllMembers(type)) {
+      if (element.kind == ElementKind.CONSTRUCTOR) {
+        constructors.add(element as ExecutableElement)
+      }
+    }
+    return constructors.size != 1 || !constructors[0].parameters.isEmpty()
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt:69]: Could use `isNotEmpty` here but that's a style choice 🤷‍♀️

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  /**
+   * Validates Interactor name's suffix.
+   *
+   * @param type
+   * @return if given type element is valid.
+   */
+  private fun validateInteractorSuffix(type: TypeElement): Boolean {
+    return if (!type.simpleName.toString().endsWith(Constants.INTERACTOR_SUFFIX)) {
+      errorReporter.reportError(String.format(Locale.getDefault(), "%s does not end in Interactor.", type))
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt:80]: Can we just use direct string interpolation with `type` for these error strings?

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  private fun validateInteractorSubclass(type: TypeElement): Boolean {
+    val interactorElement = elementUtils.getTypeElement(Interactor::class.java.name)
+    val rawElement = typesUtils.getDeclaredType(
+      interactorElement,
+      typesUtils.getWildcardType(null, null),
+      typesUtils.getWildcardType(null, null)
+    )
+    return if (!typesUtils.isSubtype(type.asType(), rawElement)) {
+      errorReporter.reportError(
+        String.format(
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt:102]: Same string interpolation comment as above

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+          Locale.getDefault(),
+          "%s is annotated with @RibInteractor but is not an " + "Interactor subclass.",
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt:104]: Also not clear to me why this is a concatenated String. 🤔

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibInteractorProcessorPipeline.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+/** The processor pipeline for [RibInteractor]  */
+open class RibInteractorProcessorPipeline(
+  processContext: ProcessContext,
+  private var interactorGenerator: Generator<InteractorAnnotatedClass>?
+) : TypeProcessorPipeline(processContext) {
+  private val annotationVerifier: AnnotationVerifier<InteractorAnnotatedClass>
+  private val builderAnnotatedClassesList: MutableList<InteractorAnnotatedClass?> = ArrayList()
+
+  /** @return the annotation type this processor pipeline deals with.
+   */
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibInteractorProcessorPipeline.kt:32]: nit - formatting of this kdoc

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibProcessor.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      return false
+    }
+    for (processorPipeline in processorPipelines) {
+      try {
+        processorPipeline.process(annotations, roundEnv)
+      } catch (e: Throwable) {
+        errorReporter?.reportError(
+          String.format(
+            Locale.getDefault(),
+            "Fatal error running %s processor: %s",
```
Comment (ekager) [inline android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibProcessor.kt:64]: Same comment about direct string interpolation

**Review by ekager — APPROVED** [review]
🔥

**Comment by jbarr21** [comment]
All nits addressed via 7943796

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/AnnotationVerifier.kt (hunk 1 above)
Comment (jbarr21) [reply android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/AnnotationVerifier.kt]: yep!

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt (hunk 2 above)
Comment (jbarr21) [reply android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt]: good call

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt (hunk 3 above)
Comment (jbarr21) [reply android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/InteractorAnnotationVerifier.kt]: yep, not sure why kotlin converter didn't do that for us

File: android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibInteractorProcessorPipeline.kt (hunk 6 above)
Comment (jbarr21) [reply android/libraries/rib-compiler-app/src/main/kotlin/com/uber/rib/compiler/RibInteractorProcessorPipeline.kt]: good catch, ktfmt will fix this for us when we can enable it via spotless

**Pull request merged**